internal/preflight/preflight.go

```diff
@@ -63,39 +65,60 @@ func (p *preflight) PreflightUpgradeCheck(ctx context.Context, pv *kmmv1beta1.Pr
 		return false, fmt.Sprintf("Failed to substitute template in kernel mapping in the module %s for kernel version %s", mod.Name, kernelVersion)
 	}
 
+	shouldBuild := module.ShouldBeBuilt(mod.Spec, *mapping)
+	shouldSign := module.ShouldBeSigned(mod.Spec, *mapping)
+
 	err = p.statusUpdater.PreflightSetVerificationStage(ctx, pv, mod.Name, kmmv1beta1.VerificationStageImage)
 	if err != nil {
 		log.Info(utils.WarnString("failed to update the stage of Module CR in preflight to image stage"), "module", mod.Name, "error", err)
 	}
 
-	imageVerified, msg := p.helper.verifyImage(ctx, mapping, mod, kernelVersion)
-	if imageVerified || (mapping.Build == nil && mod.Spec.ModuleLoader.Container.Build == nil) {
-		return imageVerified, msg
+	verified, msg := p.helper.verifyImage(ctx, mapping, mod, kernelVersion)
+	if verified {
+		shouldBuild = false
+		shouldSign = false
 	}
 
-	err = p.statusUpdater.PreflightSetVerificationStage(ctx, pv, mod.Name, kmmv1beta1.VerificationStageBuild)
-	if err != nil {
-		log.Info(utils.WarnString("failed to update the stage of Module CR in preflight to build stage"), "module", mod.Name, "error", err)
+	if shouldBuild {
+		err = p.statusUpdater.PreflightSetVerificationStage(ctx, pv, mod.Name, kmmv1beta1.VerificationStageBuild)
+		if err != nil {
+			log.Info(utils.WarnString("failed to update the stage of Module CR in preflight to build stage"), "module", mod.Name, "error", err)
+		}
+
+		verified, msg = p.helper.verifyBuild(ctx, pv, mapping, mod)
+		if !verified {
+			shouldSign = false
+		}
 	}
 
-	return p.helper.verifyBuild(ctx, pv, mapping, mod)
+	if shouldSign {
+		err = p.statusUpdater.PreflightSetVerificationStage(ctx, pv, mod.Name, kmmv1beta1.VerificationStageSign)
+		if err != nil {
+			log.Info(utils.WarnString("failed to update the stage of Module CR in preflight to sign stage"), "module", mod.Name, "error", err)
+		}
+		verified, msg = p.helper.verifySign(ctx, pv, mapping, mod)
+	}
+	return verified, msg
```

Comment:
```suggestion
	return true, fmt.Sprintf("Preflight upgrade check went successfuly")
```

Reply:
I think this whole function will be more readable if written like:
```
diff --git a/internal/preflight/preflight.go b/internal/preflight/preflight.go
index ebeea84..5623d06 100644
--- a/internal/preflight/preflight.go
+++ b/internal/preflight/preflight.go
@@ -75,8 +75,7 @@ func (p *preflight) PreflightUpgradeCheck(ctx context.Context, pv *kmmv1beta1.Pr

        verified, msg := p.helper.verifyImage(ctx, mapping, mod, kernelVersion)
        if verified {
-               shouldBuild = false
-               shouldSign = false
+               return false, msg
        }

        if shouldBuild {
@@ -85,9 +84,8 @@ func (p *preflight) PreflightUpgradeCheck(ctx context.Context, pv *kmmv1beta1.Pr
                        log.Info(utils.WarnString("failed to update the stage of Module CR in preflight to build stage"), "module", mod.Name, "error", err)
                }

-               verified, msg = p.helper.verifyBuild(ctx, pv, mapping, mod)
-               if !verified {
-                       shouldSign = false
+               if verified, msg = p.helper.verifyBuild(ctx, pv, mapping, mod); !verified {
+                       return false, msg
                }
        }

@@ -96,9 +94,11 @@ func (p *preflight) PreflightUpgradeCheck(ctx context.Context, pv *kmmv1beta1.Pr
                if err != nil {
                        log.Info(utils.WarnString("failed to update the stage of Module CR in preflight to sign stage"), "module", mod.Name, "error", err)
                }
-               verified, msg = p.helper.verifySign(ctx, pv, mapping, mod)
+               if verified, msg = p.helper.verifySign(ctx, pv, mapping, mod); !verified {
+                       return false, msg
+               }
        }
-       return verified, msg
+       return true, fmt.Sprintf("Preflight upgrade check went successfuly")
 }

```

Reply:
2 reason why not to change it:
1) verified can be false - for example verifySign failed
2) the message is constructed in each handle specifically, so we want to return it

Reply:
> verified can be false - for example verifySign failed

If `verifySign` failed then we already returned `false` earlier.

> the message is constructed in each handle specifically, so we want to return it

I am OK with return `msg` instead of `fmt.Sprintf("Preflight upgrade check went successfuly")` I have suggested.